### PR TITLE
Adding hooks to retrieve most recent decrypted and sent documents

### PR DESCRIPTION
**Pull request by haikuginger**
@bradenmacdonald, need a :+1: on this, and then I'll submit to upstream. Should the follow-up task handle submitting the branch to the EdX requirements as well as adding in the data logging, or should I submit to edx-platform as part of this task?

### CODE REVIEW

**Comment by bradenmacdonald** [comment]
@haikuginger Is there a new test for this functionality? That is important. Otherwise, looks good.

Bumping edx requirement will be part of the next task.

**Comment by bradenmacdonald** [comment]
Please also update `README.md` to document the changes in this PR.

**Comment by haikuginger** [comment]
@bradenmacdonald, those changes are made. Testing still looks good locally, although Travis is still segfaulting for unknown reasons.

File: src/onelogin/saml2/response.py [hunk 1]
```diff
@@ -549,3 +549,12 @@ def get_error(self):
         After executing a validation process, if it fails this method returns the cause
         """
         return self.__error
+
+    def get_xml_document(self):
+        """
+        If necessary, decrypt the XML response document, and return it.
+        """
```
Comment (bradenmacdonald) [inline src/onelogin/saml2/response.py:559]: Docstring needs to specify the return type:

```
        :returns: XML document
        :rtype: Element
```

Also, now that this method exists, you could use it to simplify the `__query(self, query):` method just above.

File: src/onelogin/saml2/auth.py [hunk 2]
```diff
@@ -451,3 +453,9 @@ def __build_signature(self, saml_data, relay_state, saml_type, sign_algorithm=On
 
         signature = dsig_ctx.signBinary(str(msg), sign_algorithm_transform)
         return b64encode(signature)
+
+    def get_last_response_xml(self):
+        return self.__last_response_xml
+
+    def get_last_request_xml(self):
+        return self.__last_request_xml
```
Comment (bradenmacdonald) [inline src/onelogin/saml2/auth.py]: 1. These methods currently return different types. `get_last_request_xml` returns a string as I expected, but `get_last_response_xml` is returning an XML element.
2. Please add docstrings that specify the return type:

```
        :returns: SAML request XML
        :rtype: string
```

**Comment by bradenmacdonald** [comment]
@haikuginger I think you still need a test for `get_last_response_xml`.

I used https://github.com/edx/edx-platform/pull/13298 to test this out with [TestShib](https://www.testshib.org/), and the "request" part worked fine, but the response part just logged this:

```
2016-08-23 23:33:16,123 INFO 13151 [third_party_auth.saml] saml.py:82 - SAML login testshib for IdP response. XML:
<Element {urn:oasis:names:tc:SAML:2.0:protocol}Response at 0x7f1c9c369dd0>
```

I was able to change the line in `process_response` to 

``` python
self.__last_response_xml = etree.tostring(response.get_xml_document(), pretty_print=True)
```

and then it worked fine. (Note the `pretty_print` which makes the log much  nicer)

**Comment by haikuginger** [comment]
I will go ahead and make those changes. One difference is that I'm going to put the string conversion in `get_last_request_xml`; that way, the (however negligible) processor time/memory hit to do the conversion will only happen when it's actually requested, rather than on every operation.

**Comment by haikuginger** [comment]
How's this looking for you, @bradenmacdonald?

**Comment by bradenmacdonald** [comment]
:+1:
- [x] I tested this: I used https://github.com/edx/edx-platform/pull/13298 to test this out with [TestShib](https://www.testshib.org/) and confirmed that both the request and response XML showed up in the console log
- [x] I read through the code
- [x] Includes documentation: updated readme

**Comment by haikuginger** [comment]
@bradenmacdonald - realized I forgot to update the docstrings on the internal methods. May not be a big deal, but wanted to give you a heads-up that those are now updated.

**Comment by haikuginger** [comment]
And found a test bug, now fixed. Should I merge this into our master, or just open a PR with upstream? Either way, I'm assuming we want to not delete this branch once we're done in case we end up referring to it in the `edx-master` requirements?

**Comment by bradenmacdonald** [comment]
Just open a PR with upstream for now, thanks. Yes, keep this branch open.

**Comment by bradenmacdonald** [comment]
@haikuginger Assuming (based on the PR) that the upstream review hasn't started yet, can you please squash this?

**Comment by haikuginger** [comment]
Rebased from [`ea8ee6`](https://github.com/open-craft/python-saml/commits/ea8ee66900fbd99b2d6d0858efb55bc5b407d294).

**Pull request closed without merging**